### PR TITLE
Add Revenue Report API/data-layer

### LINKED ISSUE #224: Analytics: Revenue Report API/data-layer
Child task of #97

**API Post / Schema**:  p7bje6-14c-p2

This issue encompasses creation of the `wp.data` API logic to support the needs of the Revenue Report. The p2 post linked above outlines the three requests needed to power the summary numbers, charts, and tabular data.

### PR DESCRIPTION
**Pull request by emdashcodes**
Closes #224.

This PR adds the ability to fetch revenue reports from the new v3 endpoint. It stores results under a keyed 'queries' object. Based on the designs, it seems like we might have different queries for the chart, vs the current page of the table for example, so it will be important to store multiple queries.

To Test:
* Run `npm install` followed by `npm test` and make sure all tests pass.
* Pull the latest WooCommerce v3 API from `feature/rest-api-v3-reports` (you might need to deactivate `wc-api-dev`).
* You need to deactivate and reactivate WooCommerce to get the correct summary MySQL tables.
* Uncomment the two `withSelect` lines in `client/analytics/report/revenue.js`, and move the call into the compose.
* Open the revenue report and confirm a request is made and some data is returned. You can also use the Redux console, and select 'wc-admin' from the context dropdown there. You should see actions fire.

### CODE REVIEW

**Review by psealock — COMMENTED** [review]
Nice work @justinshreve. A few small comments, otherwise looking good and tests pass.

In order to run this, I had to deactivate WooCommerce API Dev when switching to `feature/rest-api-v3-reports` branch of WooCommerce. I got a `WordPress database error` on the request. I'm not too sure whats going on, but at least I do see the error being handled.

File: client/store/reports/revenue/stats/selectors.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Returns revenue report details for a specific report query.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} query  Report query paremters
+ * @return {Object}        Report details
+ */
+function getReportRevenueStats( state, query ) {
+	const queries = get( state, 'reports.revenue.stats.queries', [] );
```
Comment (psealock) [inline client/store/reports/revenue/stats/selectors.js]: small nit, the default should be and object

**Comment by justinshreve** [comment]
> In order to run this, I had to deactivate WooCommerce API Dev when switching to feature/rest-api-v3-reports branch of WooCommerce. I got a WordPress database error on the request. I'm not too sure whats going on, but at least I do see the error being handled.

Oops, sorry about that @psealock. You need to deactivate and reactivate WooCommerce to get the correct summary tables. The upgrader code to do it between versions hasn't been written by Proton yet.

I'll update the testing directions.

**Comment by justinshreve** [comment]
Thanks for the review @psealock!

I handled your feedback suggestions in https://github.com/woocommerce/wc-admin/pull/293/commits/0f9eaa0c01996247691a74c7d436146e890afa2c. I'll also keep [this comment](https://github.com/woocommerce/wc-admin/pull/293/commits/0f9eaa0c01996247691a74c7d436146e890afa2c#r209792328) in mind when I work on #225.

**Review by psealock — APPROVED** [review]
Looking good, thanks for the updated instructions.

https://github.com/woocommerce/wc-admin/pull/287 introduced a few nav-util functions which overlap with yours, hence the merge conflict.

File: client/store/reports/revenue/stats/reducer.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import { ERROR } from 'store/constants';
+
+const DEFAULT_STATE = {
+	queries: {},
+};
+
+export default function reportRevenueStatsReducer( state = DEFAULT_STATE, action ) {
+	if ( 'SET_REPORT_REVENUE_STATS' === action.type ) {
+		const prevQueries = get( state, 'queries', {} );
+		const query = JSON.stringify( action.query );
```
Comment (psealock) [inline client/store/reports/revenue/stats/reducer.js]: Oh, one more small thing, you'll need to guarantee equal objects are serialised with properties in the same order:

```js
const query = JSON.stringify( action.query, Object.keys( action.query ).sort() );
```

**Pull request merged**